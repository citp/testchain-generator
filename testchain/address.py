import hashlib

from bitcointx.core import Hash160
from bitcointx.core.script import CScript, OP_CHECKSIG, OP_0
from bitcointx.wallet import CBitcoinSecret, P2PKHBitcoinAddress, P2SHBitcoinAddress, P2WSHBitcoinAddress, \
    P2WPKHBitcoinAddress

COINBASE_ADDRESS = "mjTkW3DjgyZck4KbiRusZsqTgaYTxdSz6z"
COINBASE_KEY = "cVpF924EspNh8KjYsfhgY96mmxvT6DgdWiTYMtMjuM74hJaU5psW"
UNSPENDABLE_ADDRESS = "mt7L7Qh8R5xwGjpLPJMv7tbjbQPVPXjtKJ"


class UnsupportedAddressTypeError(Exception):
    """We do not support the address type"""


class Address(object):
    """
    Stores information related to addresses such as keys or the most recent output pointer.
    """
    def __init__(self, key: CBitcoinSecret, address_type: str = "p2pkh"):
        self.key_index = None
        self.type = address_type
        self.key = key
        if address_type == "p2pkh":
            self.address = P2PKHBitcoinAddress.from_pubkey(self.key.pub)
        elif address_type == "p2sh":
            self.address = P2SHBitcoinAddress.from_redeemScript(CScript([self.key.pub, OP_CHECKSIG]))
        elif address_type == "p2wpkh":
            key_hash = Hash160(self.key.pub)
            script_pub_key = CScript([OP_0, key_hash])
            self.address = P2WPKHBitcoinAddress.from_scriptPubKey(script_pub_key)
        elif address_type == "p2wsh":
            self.witness_program = CScript([self.key.pub, OP_CHECKSIG])
            script_hash = hashlib.sha256(self.witness_program).digest()
            script_pub_key = CScript([OP_0, script_hash])
            self.address = P2WSHBitcoinAddress.from_scriptPubKey(script_pub_key)
        else:
            raise UnsupportedAddressTypeError()
        self.value = 0
        self.txid = None
        self.vout = None

    @staticmethod
    def compute_key(key_index: int) -> CBitcoinSecret:
        """
        Deterministically generate a key from an index.
        """
        h = hashlib.sha256(bytes(key_index)).digest()
        return CBitcoinSecret.from_secret_bytes(h)

    @classmethod
    def from_key_index(cls, key_index, address_type: str = "p2pkh"):
        """
        Creates an address object based on a key index.
        """
        c = cls(cls.compute_key(key_index), address_type)
        c.key_index = key_index
        return c
