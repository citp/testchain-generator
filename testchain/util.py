class Coin(object):
    COIN = 1e8

    def __init__(self, val):
        self.val = val

    @classmethod
    def from_satoshi(cls, value):
        return cls(value / cls.COIN)

    def satoshi(self):
        return int(round(self.val * self.COIN, 0))

    def bitcoin(self):
        return round(self.val, 8)

    def __eq__(self, other):
        if type(other) == Coin:
            return self.satoshi() == other.satoshi()
        else:
            return self.satoshi() == other

    def __ne__(self, other):
        if type(other) == Coin:
            return self.satoshi() != other.satoshi()
        else:
            return self.satoshi() != other

    def __le__(self, other):
        if type(other) == Coin:
            return self.satoshi() <= other.satoshi()
        else:
            return self.satoshi() <= other

    def __ge__(self, other):
        if type(other) == Coin:
            return self.satoshi() >= other.satoshi()
        else:
            return self.satoshi() >= other

    def __lt__(self, other):
        if type(other) == Coin:
            return self.satoshi() < other.satoshi()
        else:
            return self.satoshi() < other

    def __gt__(self, other):
        if type(other) == Coin:
            return self.satoshi() > other.satoshi()
        else:
            return self.satoshi() > other

    def __add__(self, other):
        if type(other) == Coin:
            return self.satoshi() + other.satoshi()
        else:
            return self.satoshi() + other

    def __sub__(self, other):
        if type(other) == Coin:
            return self.satoshi() - other.satoshi()
        else:
            return self.satoshi() - other

    def __mul__(self, other):
        if type(other) == Coin:
            return self.satoshi() * other.satoshi()
        else:
            return self.satoshi() * other

    def __div__(self, other):
        if type(other) == Coin:
            return self.satoshi() / other.satoshi()
        else:
            return self.satoshi() / other

    def __str__(self):
        return "Coin[{}]".format(self.satoshi())

    def __repr__(self):
        return "Coin[{}]".format(self.satoshi())
