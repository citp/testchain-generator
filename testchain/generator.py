from typing import List, Callable, Dict

from logging import Logger
import bitcoin.rpc
from bitcoin.core import CMutableTxIn, CMutableTxOut, CMutableTransaction, COutPoint, CTxInWitness, CTxWitness, b2lx
from bitcoin.core.script import CScript, CScriptWitness, OP_CHECKSIG, SignatureHash, SIGHASH_ALL, SIGVERSION_WITNESS_V0
from bitcoin.wallet import CBitcoinSecret

from testchain.address import Address, UnsupportedAddressTypeError, COINBASE_KEY, COINBASE_ADDRESS, UNSPENDABLE_ADDRESS
from testchain.util import Coin


class NoAddressError(Exception):
    """Raised when requesting an address, but none has been created yet."""


class Generator(object):
    addresses: List[Address]

    def run(self):
        raise NotImplementedError

    def __init__(self, proxy: bitcoin.rpc.Proxy, chain, log: Logger, stored_hashes: Dict, offset: int,
                 next_timestamp: Callable[[], int]):
        self.proxy = proxy
        self.chain = chain
        self.log = log
        self.addresses = []
        self.offset = offset
        self.address_cursor = -1
        self.stored_hashes = stored_hashes
        self.fee = 0.0001
        self._next_timestamp = next_timestamp
        self.segwit = self.chain == "btc"

    def log_value(self, k, v):
        self.stored_hashes[k] = v

    def current_block_reward(self):
        """
        Returns the current reward the miner gets for each block.
        In regtest mode, the reward is halved after the first 150 blocks.
        """
        return 50 if self.proxy.getblockcount() < 150 else 25

    def _advance_time(self):
        """
        Advance time by increasing the mocktime of the Bitcoin node
        """
        timestamp = self._next_timestamp()
        self.proxy.call("setmocktime", timestamp)
        self.log.debug("Set time to {}".format(timestamp))

    def generate_block(self, n: int = 1, spendable_coinbase=True):
        self._advance_time()

        if spendable_coinbase:
            address = COINBASE_ADDRESS
        else:
            address = UNSPENDABLE_ADDRESS

        block_hashes = self.proxy.call("generatetoaddress", n, address)
        if len(block_hashes) == 1:
            self.log.debug("Mined block: {}".format([x for x in block_hashes][0]))
        else:
            self.log.debug("Mined blocks: {}".format([x for x in block_hashes]))
        return block_hashes

    def next_address(self, address_type="p2pkh") -> Address:
        self.address_cursor += 1
        self.addresses.append(Address.from_key_index(self.address_cursor + self.offset, address_type))
        return self.addresses[self.address_cursor]

    def current_address(self) -> Address:
        if self.address_cursor < 0:
            raise NoAddressError("No addresses exist yet. Create one using `next_address()`")
        return self.addresses[self.address_cursor]

    def fund_address(self, address: Address, value: float):
        """
        Sends money to an address using a UTXO of the mining address
        :param address: the address that will receive the coins
        :param value: the value that the address will receive
        :return: the transaction id and
        """
        key = CBitcoinSecret(COINBASE_KEY)
        coinbase_addr = Address(key)

        unspents = self.proxy.listunspent(minconf=100, addrs=[coinbase_addr.address])
        unspents.sort(key=lambda x: x['confirmations'], reverse=True)
        oldest_utxo = next(x for x in unspents if Coin.from_satoshi(x['amount']).bitcoin() >= value + self.fee)

        coinbase_addr.txid = oldest_utxo['outpoint'].hash
        coinbase_addr.vout = oldest_utxo['outpoint'].n

        coinbase_addr.value = Coin.from_satoshi(oldest_utxo['amount']).bitcoin()
        address.value = value

        # change address is always the second output
        txid = self.create_transaction([coinbase_addr], [address, coinbase_addr],
                                       values=[value, coinbase_addr.value - self.fee - value])

        self.log.debug("Address: {}".format(address.address))
        self.log.debug("Tx: {}".format(txid))

        return txid, address.vout

    def create_transaction(self, sources: List[Address], recipients: List[Address], values=[], n_locktime=0,
                           n_sequence=0xffffffff):
        """
        Creates a transaction using `sources` as inputs and `recipients` as outputs.
        :param sources: list of addresses to spend from
        :param recipients: list of addresses to send to
        :param n_locktime: locktime value of the transaction
        :param n_sequence: sequence no for inputs
        :return: the TXID of the transaction as returned by the proxy
        """
        tx = self._create_transaction(sources, recipients, values, n_locktime=n_locktime, n_sequence=n_sequence)
        return self._send_transaction(tx, recipients)

    def _create_transaction(self, sources: List[Address], recipients: List[Address], values, n_locktime, n_sequence):
        if not values:
            values = [recipient.value for recipient in recipients]
        tx_ins = [CMutableTxIn(COutPoint(source.txid, source.vout), nSequence=n_sequence) for source in sources]
        tx_outs = []

        cnt = 0
        for recipient, value in zip(recipients, values):
            if recipient.value == 0:
                self.log.warning("Creating output with 0 BTC")
            recipient.vout = cnt

            tx_outs.append(CMutableTxOut(Coin(value).satoshi(), recipient.address.to_scriptPubKey()))
            cnt += 1

        tx = CMutableTransaction(tx_ins, tx_outs, nLockTime=n_locktime)

        in_idx = 0
        witnesses = []
        for txin, source in zip(tx_ins, sources):
            key = source.key

            if source.type == 'p2pkh':
                script = source.address.to_redeemScript()
            elif source.type == 'p2sh':
                script = CScript([key.pub, OP_CHECKSIG])
            elif source.type == 'p2wpkh':
                script = source.address.to_redeemScript()
            elif source.type == 'p2wsh':
                script = source.witness_program
            else:
                raise UnsupportedAddressTypeError()

            # Create signature
            amount = Coin(source.value).satoshi()
            sig = self._sign(script, tx, in_idx, amount, key, source.type)

            # Add signature to input or witness
            if source.type == 'p2pkh':
                txin.scriptSig = CScript([sig, key.pub])
                witnesses.append(CTxInWitness())
            elif source.type == 'p2sh':
                txin.scriptSig = CScript([sig, script])
                witnesses.append(CTxInWitness())
            elif source.type == 'p2wpkh':
                txin.scriptSig = CScript()
                witnesses.append(CTxInWitness(CScriptWitness([sig, key.pub])))
            elif source.type == 'p2wsh':
                txin.scriptSig = CScript()
                witnesses.append(CTxInWitness(CScriptWitness([sig, script])))
            in_idx += 1

        tx.wit = CTxWitness(witnesses)
        return tx

    def _send_transaction(self, tx: CMutableTransaction, recipients: List[Address]):
        txid = self.proxy.sendrawtransaction(tx)
        for rec in recipients:
            rec.txid = txid
        return b2lx(txid)

    def _sign(self, script, tx, in_idx, amount, key, script_type="p2pkh"):
        if self.chain == "bch":
            sighash = SignatureHash(script, tx, in_idx, SIGHASH_ALL | 0x40, amount=amount,
                                    sigversion=SIGVERSION_WITNESS_V0)
            sig = key.sign(sighash) + bytes([SIGHASH_ALL | 0x40])
        else:
            if script_type == 'p2wpkh' or script_type == 'p2wsh':
                sighash = SignatureHash(script, tx, in_idx, SIGHASH_ALL, amount=amount,
                                        sigversion=SIGVERSION_WITNESS_V0)
            else:
                sighash = SignatureHash(script, tx, in_idx, SIGHASH_ALL)
            sig = key.sign(sighash) + bytes([SIGHASH_ALL])
        return sig
