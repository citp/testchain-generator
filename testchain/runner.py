import atexit
import json
import logging
import os
import shutil
import subprocess
import tempfile
from time import sleep
from typing import List, Type

import bitcoin
import bitcoin.rpc
from bitcoin.core.key import use_libsecp256k1_for_signing
from testchain.generator import Generator
from testchain.address import COINBASE_KEY

LOG_LEVEL = logging.INFO
bitcoin.SelectParams('regtest')

use_libsecp256k1_for_signing(True)  # for deterministic coinbase transactions and signatures


class Runner(object):
    motif_generators: List[Generator]

    def __init__(self, output_dir, chain, exec):
        self.chain = chain
        self.exec = exec
        self.current_time = 1535760000
        self.prev_block = None
        self.motif_generators = []
        self.kv = {}
        self.output_dir = os.path.join(output_dir, '')
        self._setup_logger()
        self._setup_bitcoind()
        self.proxy = bitcoin.rpc.Proxy(btc_conf_file=self._conf_file())
        self.proxy.call("importprivkey", COINBASE_KEY)

    def _setup_logger(self):
        self.log = logging.getLogger(__name__)
        self.log.setLevel(LOG_LEVEL)
        ch = logging.StreamHandler()
        ch.setLevel(LOG_LEVEL)
        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        ch.setFormatter(formatter)
        self.log.addHandler(ch)

    def _setup_bitcoind(self):
        self.tempdir = tempfile.TemporaryDirectory()
        self.log.info("bitcoind datadir: {}".format(self.tempdir.name))

        # copy conf file to temp dir
        shutil.copy("bitcoin.conf", self.tempdir.name)

        # launch bitcoind
        params = [self.exec, "-rpcport=18443", "-datadir={}".format(self.tempdir.name),
                  "-mocktime={}".format(self.current_time)]

        # Disable Bitcoin Cash specific address format (breaks Python library)
        # Enable CTOR
        if self.chain == "bch":
            params += ["-usecashaddr=0", "-magneticanomalyactivationtime=0"]
        self.proc = subprocess.Popen(params, stdout=subprocess.DEVNULL)

        # kill process when generator is done
        atexit.register(self._terminate)

        self.log.info("Waiting 5 seconds for bitcoind to start")
        sleep(5)

    def _conf_file(self):
        return "{}/bitcoin.conf".format(self.tempdir.name)

    def _terminate(self):
        """
        Kills the bitcoind process
        """
        self.proc.terminate()
        self.log.info("Waiting 5 seconds for bitcoind to quit")
        sleep(5)

    def next_timestamp(self):
        self.current_time += 600
        return self.current_time

    def export_address_counts(self):
        self._address_sanity_check()
        counts = {"p2pkh": 2, "p2wpkh": 0, "p2sh": 0, "p2wsh": 0}  # coinbase addresses
        for g in self.motif_generators:
            for addr in g.addresses:
                counts[addr.type] += 1
        self.kv["p2pkh_address_count"] = counts["p2pkh"]
        self.kv["p2wpkh_address_count"] = counts["p2wpkh"]
        self.kv["p2sh_address_count"] = counts["p2sh"]
        self.kv["p2wsh_address_count"] = counts["p2wsh"]

    def _address_sanity_check(self):
        total_addresses = 0
        unique_addresses = set()
        for g in self.motif_generators:
            key_indizes = [x.key_index for x in g.addresses]
            unique_addresses |= set(key_indizes)
            total_addresses += len(key_indizes)
        if len(unique_addresses) != total_addresses:
            self.log.warning("Addresses are not unique.")

    def copy_blk_file(self, truncate_file=True):
        """
        Copies the first blk file from the regtest directory to the output directory
        :param truncate_file: Whether the final block file should be truncated. Works with BlockSci, but may not work
        when using other parsers.
        """
        blk_destination = self.output_dir + self.chain + "/regtest/blocks/"
        self.log.info("Copying blk00000.dat to {}".format(blk_destination))
        if not os.path.exists(blk_destination):
            os.makedirs(blk_destination)
        source = "{}/regtest/blocks/blk00000.dat".format(self.tempdir.name)

        if truncate_file:
            with open(source, "rb") as f:
                with open(blk_destination + "blk00000.dat", "wb") as dest:
                    counter = 0
                    while True:
                        bts = f.read(16)
                        if not bts or counter == 16:
                            break
                        if bts == b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00':
                            counter += 1
                        else:
                            counter = 0
                        dest.write(bts)
        else:
            shutil.copy(source, blk_destination)

    def persist_hashes(self):
        """
        Dumps hashes into JSON file.
        """
        # kv = {}
        # for g in self.motif_generators:
        #     kv = {**kv, **g.stored_hashes}

        self.log.info("Writing hashes to file output.json")
        self.log.debug(self.kv)
        dest_dir = self.output_dir + self.chain + "/"
        if not os.path.exists(dest_dir):
            os.mkdir(dest_dir)
        with open(dest_dir + "output.json", "w") as f:
            json.dump(self.kv, f, indent=4)

    def add_generator(self, generator: Type[Generator]):
        gen = generator(self.proxy, self.chain, self.log, self.kv, (len(self.motif_generators) + 1) * 10000,
                        self.next_timestamp)
        self.log.debug("Magic No: {}".format(gen.offset))
        self.motif_generators.append(gen)

    def run(self):
        for g in self.motif_generators:
            g.run()
        self._address_sanity_check()
        self.copy_blk_file()
        self.persist_hashes()
