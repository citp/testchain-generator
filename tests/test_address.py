from testchain.address import Address


def test_p2pkh_address(regtest):
    addr = Address.from_key_index(0, address_type="p2pkh")
    assert 0 == addr.key_index
    assert "p2pkh" == addr.type
    assert 0 == addr.value

    address_regression(addr, regtest)


def test_p2sh_address(regtest):
    addr = Address.from_key_index(1, address_type="p2sh")
    assert 1 == addr.key_index
    assert "p2sh" == addr.type
    assert 0 == addr.value

    address_regression(addr, regtest)


def test_p2wpkh_address(regtest):
    addr = Address.from_key_index(2, address_type="p2wpkh")
    assert 2 == addr.key_index
    assert "p2wpkh" == addr.type
    assert 0 == addr.value

    address_regression(addr, regtest)


def test_p2wsh_address(regtest):
    addr = Address.from_key_index(3, address_type="p2wsh")
    assert 3 == addr.key_index
    assert "p2wsh" == addr.type
    assert 0 == addr.value

    address_regression(addr, regtest)


def address_regression(addr, regtest):
    print(addr.address, file=regtest)
    print(addr.key, file=regtest)
    print(type(addr.address), file=regtest)
