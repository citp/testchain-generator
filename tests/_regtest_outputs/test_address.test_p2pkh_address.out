muZpTpBYhxmRFuCjLc7C6BBDF32C8XVJUi
cVDJUtDjdaM25yNVVDLLX3hcHUfth4c7tY3rSc4hy9e8ibtCuj6G
<class 'bitcointx.wallet.P2PKHBitcoinAddress'>
